aeron-driver/src/main/c/aeron_retransmit_handler.c

```diff
@@ -139,41 +143,87 @@ int aeron_retransmit_handler_process_timeouts(
 
-                num_active_actions--;
+        if (AERON_RETRANSMIT_ACTION_STATE_DELAYED == action->state)
+        {
+            if (now_ns > action->expiry_ns)
+            {
+                result = resend(resend_clientd, action->term_id, action->term_offset, action->length);
+                action->state = AERON_RETRANSMIT_ACTION_STATE_LINGERING;
+                action->expiry_ns = now_ns + (int64_t)handler->linger_timeout_ns;
+                result++;
             }
-            else if (AERON_RETRANSMIT_ACTION_STATE_DELAYED == action->state)
+        }
+        else if (AERON_RETRANSMIT_ACTION_STATE_LINGERING == action->state)
+        {
+            if (now_ns > action->expiry_ns)
             {
-                if (now_ns > action->expiry_ns)
+                aeron_retransmit_handler_remove_retransmit(handler, action);
+                result++;
+            }
+        }
+    }
+
+    return result;
+}
+
+int aeron_retransmit_handler_scan_for_available_retransmit(
+    aeron_retransmit_handler_t *handler,
+    int32_t term_id,
+    int32_t term_offset,
+    size_t length,
+    aeron_retransmit_action_t **actionp)
+{
+   if (0 == handler->active_retransmits)
+   {
+       *actionp = aeron_retransmit_handler_add_retransmit(handler, &handler->retransmit_action_pool[0]);
+       return 0;
+   }
+
+    aeron_retransmit_action_t *available_action = NULL;
+    for (size_t i = 0; i < AERON_RETRANSMIT_HANDLER_MAX_RETRANSMITS; i++)
+    {
+        aeron_retransmit_action_t *action = &handler->retransmit_action_pool[i];
+
+        switch (action->state)
+        {
+            case AERON_RETRANSMIT_ACTION_STATE_INACTIVE:
+                if (NULL == available_action)
                 {
-                    result = resend(resend_clientd, action->term_id, action->term_offset, action->length);
-                    action->state = AERON_RETRANSMIT_ACTION_STATE_LINGERING;
-                    action->expiry_ns = now_ns + (int64_t)handler->linger_timeout_ns;
-                    result++;
+                    available_action = action;
                 }
+                break;
 
-                num_active_actions--;
-            }
+            case AERON_RETRANSMIT_ACTION_STATE_DELAYED:
+            case AERON_RETRANSMIT_ACTION_STATE_LINGERING:
+                if (action->term_id == term_id &&
+                    action->term_offset <= term_offset &&
+                    term_offset + length <= action->term_offset + action->length)
+                {
+                    *actionp = NULL;
+                    return 0;
+                }
+                break;
         }
     }
 
-    return result;
+    if (NULL != available_action)
+    {
+        *actionp = aeron_retransmit_handler_add_retransmit(handler, available_action);
+        return 0;
+    }
+
+    // it shouldn't be possible to get here...
+    AERON_SET_ERR(EINVAL, "%s", "no available retransmit actions - active retransmit counter appears to be out of sync");
```

Comment:
Use the same error message that we have in the Java driver.  Makes the behaviour more consistent.